File: frontend/src/shared/utils/formato.ts
/** Fecha local YYYY-MM-DD (evita desfase UTC con toISOString). */
export function fechaHoyLocal(): string {
  const d = new Date();
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/** "06:00:00" → "6:00 AM" (formato Nicaragua, 12 horas) */
export function formatearHora(hora: string): string {
  return formatearHoraNicaragua(hora);
}

/** Convierte hora 24h (HH:mm o HH:mm:ss) a formato 12h nicaragüense. */
export function formatearHoraNicaragua(hora: string): string {
  const parts = hora.trim().split(':');
  const h24 = parseInt(parts[0] ?? '0', 10);
  const min = (parts[1] ?? '00').slice(0, 2).padStart(2, '0');

  if (Number.isNaN(h24)) return hora;

  const esPm = h24 >= 12;
  let h12 = h24 % 12;
  if (h12 === 0) h12 = 12;

  return `${h12}:${min} ${esPm ? 'PM' : 'AM'}`;
}

/** Parsea "1:30 PM" → "13:30" para enviar al backend */
export function horaNicaraguaA24h(hora12: string): string {
  const match = hora12.trim().match(/^(\d{1,2}):(\d{2})\s*(AM|PM)$/i);
  if (!match) return hora12;

  let h = parseInt(match[1], 10);
  const m = match[2];
  const ampm = match[3].toUpperCase();

  if (ampm === 'AM') {
    if (h === 12) h = 0;
  } else if (h !== 12) {
    h += 12;
  }

  return `${String(h).padStart(2, '0')}:${m}`;
}

/** Acepta "6:00 AM" o "06:00" y devuelve "HH:mm" para la API. */
export function horaParaBackend(hora: string): string {
  const t = hora.trim();
  if (/^\d{1,2}:\d{2}\s*(AM|PM)$/i.test(t)) {
    return horaNicaraguaA24h(t);
  }
  const match = t.match(/^(\d{1,2}):(\d{2})(?::(\d{2}))?$/);
  if (match) {
    return `${String(parseInt(match[1], 10)).padStart(2, '0')}:${match[2]}`;
  }
  return t;
}

export type Meridiano = 'AM' | 'PM';

export interface HoraSalidaNicaragua {
  hora: number;
  minuto: string;
  ampm: Meridiano;
}

export const HORA_SALIDA_DEFAULT: HoraSalidaNicaragua = { hora: 6, minuto: '00', ampm: 'AM' };

/** Descompone "06:00" o "18:30:00" en hora 12h + AM/PM. */
export function descomponerHora24(hora24: string): HoraSalidaNicaragua {
  const parts = hora24.trim().split(':');
  const h24 = parseInt(parts[0] ?? '0', 10);
  const min = (parts[1] ?? '00').slice(0, 2).padStart(2, '0');
  if (Number.isNaN(h24) || h24 < 0 || h24 > 23) return HORA_SALIDA_DEFAULT;

  const esPm = h24 >= 12;
  let h12 = h24 % 12;
  if (h12 === 0) h12 = 12;

  return { hora: h12, minuto: min, ampm: esPm ? 'PM' : 'AM' };
}

export function componerHoraNicaragua(v: HoraSalidaNicaragua): string {
  return `${v.hora}:${v.minuto} ${v.ampm}`;
}

export function componerHoraBackend(v: HoraSalidaNicaragua): string {
  return horaNicaraguaA24h(componerHoraNicaragua(v));
}

export function formatearCordobas(monto: number): string {
  return `C$ ${monto.toLocaleString('es-NI', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

export const ETIQUETAS_ESTADO_ASIENTO: Record<string, string> = {
  DISPONIBLE: 'Disponible',
  VENDIDO: 'Vendido',
  CANCELADO: 'Cancelado',
  RESERVADO_EXCEPCIONAL: 'Reserva excepcional',
};

export type PosicionAsiento =
  | 'VENTANA'
  | 'PASILLO'
  | 'TRASERA_1'
  | 'TRASERA_2'
  | 'TRASERA_3'
  | 'TRASERA_4'
  | 'TRASERA_5';

export const ETIQUETAS_POSICION_ASIENTO: Record<string, string> = {
  VENTANA: 'Ventana',
  PASILLO: 'Pasillo',
  TRASERA_1: 'Trasera',
  TRASERA_2: 'Trasera',
  TRASERA_3: 'Trasera',
  TRASERA_4: 'Trasera',
  TRASERA_5: 'Trasera',
};

export function etiquetaPosicionAsiento(posicion: string): string {
  if (posicion.startsWith('TRASERA')) return 'Fila trasera';
  return ETIQUETAS_POSICION_ASIENTO[posicion] ?? posicion;
}

export function esAsientoVentana(posicion: string): boolean {
  return posicion === 'VENTANA';
}

export function esAsientoTrasera(posicion: string): boolean {
  return posicion.startsWith('TRASERA');
}

export const ETIQUETAS_ESTADO_VIAJE: Record<string, string> = {
  PROGRAMADO: 'Programado',
  EN_CURSO: 'En curso',
  COMPLETADO: 'Completado',
  CANCELADO: 'Cancelado',
};

File: mobile/lib/core/utils/formato.dart
import 'package:intl/intl.dart';

String formatearCordobas(double monto) {
  final fmt = NumberFormat.currency(locale: 'es_NI', symbol: 'C\$', decimalDigits: 2);
  return fmt.format(monto);
}

/// "06:00:00" → "6:00 AM" (formato Nicaragua, 12 horas).
String formatearHora(String hora) {
  final parts = hora.trim().split(':');
  if (parts.isEmpty) return hora;

  final h24 = int.tryParse(parts[0]) ?? -1;
  if (h24 < 0 || h24 > 23) return hora;

  final minRaw = parts.length > 1 ? parts[1] : '00';
  final min = minRaw.length >= 2 ? minRaw.substring(0, 2) : minRaw.padLeft(2, '0');

  final esPm = h24 >= 12;
  var h12 = h24 % 12;
  if (h12 == 0) h12 = 12;

  return '$h12:$min ${esPm ? 'PM' : 'AM'}';
}

/// Parsea "1:30 PM" → "13:30" para enviar al backend.
String horaNicaraguaA24h(String hora12) {
  final match = RegExp(r'^(\d{1,2}):(\d{2})\s*(AM|PM)$', caseSensitive: false).firstMatch(hora12.trim());
  if (match == null) return hora12;

  var h = int.parse(match.group(1)!);
  final m = match.group(2)!;
  final ampm = match.group(3)!.toUpperCase();

  if (ampm == 'AM') {
    if (h == 12) h = 0;
  } else if (h != 12) {
    h += 12;
  }

  return '${h.toString().padLeft(2, '0')}:$m';
}

/// Acepta "6:00 AM" o "06:00" y devuelve "HH:mm" para la API.
String horaParaBackend(String hora) {
  final t = hora.trim();
  if (RegExp(r'^\d{1,2}:\d{2}\s*(AM|PM)$', caseSensitive: false).hasMatch(t)) {
    return horaNicaraguaA24h(t);
  }
  final match = RegExp(r'^(\d{1,2}):(\d{2})(?::(\d{2}))?$').firstMatch(t);
  if (match != null) {
    final h = int.parse(match.group(1)!);
    final min = match.group(2)!;
    return '${h.toString().padLeft(2, '0')}:$min';
  }
  return t;
}

class HoraSalidaNicaragua {
  const HoraSalidaNicaragua({required this.hora, required this.minuto, required this.ampm});

  final int hora;
  final String minuto;
  final String ampm;

  static const inicial = HoraSalidaNicaragua(hora: 6, minuto: '00', ampm: 'AM');

  factory HoraSalidaNicaragua.desdeBackend(String hora24) => descomponerHora24(hora24);

  String toNicaragua() => '$hora:$minuto $ampm';

  String toBackend() => horaParaBackend(toNicaragua());
}

HoraSalidaNicaragua descomponerHora24(String hora24) {
  final parts = hora24.trim().split(':');
  if (parts.isEmpty) return HoraSalidaNicaragua.inicial;

  final h24 = int.tryParse(parts[0]) ?? -1;
  if (h24 < 0 || h24 > 23) return HoraSalidaNicaragua.inicial;

  final minRaw = parts.length > 1 ? parts[1] : '00';
  final min = minRaw.length >= 2 ? minRaw.substring(0, 2) : minRaw.padLeft(2, '0');

  final esPm = h24 >= 12;
  var h12 = h24 % 12;
  if (h12 == 0) h12 = 12;

  return HoraSalidaNicaragua(hora: h12, minuto: min, ampm: esPm ? 'PM' : 'AM');
}

String fechaHoyIso() {
  return DateFormat('yyyy-MM-dd').format(DateTime.now());
}

String formatearFechaCorta(String iso) {
  final d = DateTime.tryParse(iso);
  if (d == null) return iso;
  const meses = ['ene', 'feb', 'mar', 'abr', 'may', 'jun', 'jul', 'ago', 'sep', 'oct', 'nov', 'dic'];
  return '${d.day} ${meses[d.month - 1]}';
}

String etiquetaEstadoAsiento(String estado) {
  switch (estado) {
    case 'DISPONIBLE':
      return 'Disponible';
    case 'VENDIDO':
      return 'Vendido';
    case 'CANCELADO':
      return 'Cancelado';
    case 'RESERVADO_EXCEPCIONAL':
      return 'Reservado';
    default:
      return estado;
  }
}
